diff --git a/config/stack_config.h b/config/stack_config.h
index 5490719..532fcaa 100644
--- a/config/stack_config.h
+++ b/config/stack_config.h
@@ -18,7 +18,7 @@
 #define DEBUG_ISO_SERVER 0
 #define DEBUG_ISO_CLIENT 0
 #define DEBUG_IED_SERVER 0
-#define DEBUG_IED_CLIENT 0
+#define DEBUG_IED_CLIENT 0
 #define DEBUG_MMS_CLIENT 0
 #define DEBUG_MMS_SERVER 0
 #define DEBUG_GOOSE_SUBSCRIBER 0
@@ -37,7 +37,7 @@
  * 0 ==> server runs in multi-threaded mode (one thread for each connection and
  * one server background thread )
  */
-#define CONFIG_MMS_SINGLE_THREADED 0
+#define CONFIG_MMS_SINGLE_THREADED 1
 
 #if (WITH_MBEDTLS == 1)
 #define CONFIG_MMS_SUPPORT_TLS 1
@@ -74,10 +74,10 @@
 //#define CONFIG_ETHERNET_INTERFACE_ID "en0"  // OS X uses enX in place of ethX as ethernet NIC names.
 
 /* Set to 1 to include GOOSE support in the build. Otherwise set to 0 */
-#define CONFIG_INCLUDE_GOOSE_SUPPORT 1
+#define CONFIG_INCLUDE_GOOSE_SUPPORT 0
 
 /* Set to 1 to include Sampled Values support in the build. Otherwise set to 0 */
-#define CONFIG_IEC61850_SAMPLED_VALUES_SUPPORT 1
+#define CONFIG_IEC61850_SAMPLED_VALUES_SUPPORT 0
 
 /* Set to 1 to compile for edition 1 server - default is 0 to compile for edition 2 */
 #define CONFIG_IEC61850_EDITION_1 0
